<a href="https://colab.research.google.com/github/IsabellaHerrerae/Anal-tica-/blob/main/Reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

0. Descargar bibliotecas

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier

1. Traer las bases de datos

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

nxl='/content/drive/MyDrive/Colab Notebooks/2. BD1_Travel_Data_int.xlsx'
XDB=pd.read_excel(nxl, sheet_name=0)
XDB=XDB.dropna()
XDB.head(100)

XD=np.array(XDB[['College','FoodSpend','Income','TravelSpend']]) #Estos son los datos de entrada
yd=np.array(XDB['CreditCard']) #Esta es la variable de salida y el dato que quiro que el modelo pronostique

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


2. Creación del modelo (Decision Tree)

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4) #4 niveles o variables de entrada
mar.fit(XD,yd)
ydp=mar.predict(XD)

from sklearn.metrics import confusion_matrix
cm=confusion_matrix(yd,ydp)
print(cm)

#Para evalaur el modelo usamos La matriz de confusión, y esta es: 0,0=Verdaderos negavtivos - 1,1=Verdaderos postivos - 0,1=Falsos negativos y 1,0=Falso Positivo
VN=cm[0,0]
print("Verdaderos negativos =",VN)
FP=cm[0,1]
print("Falsos positivos =",FP)
FN=cm[1,0]
print("Falsos negativos =",FN)
VP=cm[1,1]
print("Verdaderos positivos =",VP)

Ex=(VN+VP)/(VN+VP+FN+FP) #Cuantos datos o que porcentaje de estos, fueron pronosticados correctamente
print("La exactitud alcanzada por el modelo es:",Ex)
Se=VP/(FN+VP) # Sensiilidad = Como le fue con los positivos
print("La sensibilidad alcanzada por el modelo es:",Se)
Sp=VN/(VN+FP) # Especificidad = Como le fue con los negativos
print("La especificidad alcanzada por el modelo es:",Sp)

[[334   9]
 [ 42 115]]
Verdaderos negativos = 334
Falsos positivos = 9
Falsos negativos = 42
Verdaderos positivos = 115
La exactitud alcanzada por el modelo es: 0.898
La sensibilidad alcanzada por el modelo es: 0.732484076433121
La especificidad alcanzada por el modelo es: 0.9737609329446064


3, Encontrar el árbol

In [ ]:
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['College','FoodSpend','Income','TravelSpend']
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol.png')

True

4. Encontrar la probabilidad de gastos de los nuevos clientes
la base de datos Travel_Plan_Score, Cuál es la probabilidad de que el primer cliente utilizará la tarjeta de crédito para el pago de los gastos de viaje y de los alimentos en sus planes de viajar el próximo año, y Cuál es la probabilidad para el segundo cliente.

In [ ]:

XDB2=pd.read_excel(nxl, sheet_name=1)
XDB2=XDB2.dropna()
XDB2.head(100)

XD2=np.array(XDB2[['College','FoodSpend','Income','TravelSpend']]) #Estos son los datos de entrada
print(XD2)

yd2=mar.predict(XD2)
print(yd2)
print(np.sum(yd2))
print(len(yd2)-np.sum(yd2))
print(len(yd2))

[[0.00000e+00 2.89290e+03 6.59820e+04 2.15160e+03]
 [1.00000e+00 6.01766e+03 5.39860e+04 2.00250e+03]]
[0 0]
0
2
2


5.
El árbol de decisión descrito utiliza las variables TravelSpend (gasto en viajes), Income (ingreso) y FoodSpend (gasto en alimentos) para predecir si un cliente usará su tarjeta de crédito para pagar gastos de viaje y alimentos. El modelo tiene una exactitud general del 89.8%, lo que significa que clasifica correctamente la mayoría de los casos y lo hace un buen modelo. Además de esto, presenta una sensibilidad del 73.2%, indicando que identifica correctamente el 73.2% de los casos positivos, y una especificidad del 97.3%, mostrando que identifica correctamente el 97.3% de los casos negativos.

Para los nuevos clientes, el modelo predice que el primer cliente tiene un 100% de probabilidad de usar la tarjeta de crédito, ya que pertenece a un nodo puro donde todos los datos (108 casos) corresponden a clientes que usan la tarjeta. Este cliente sigue la regla de negocio: Si TravelSpend es ≥ 1556 e Income es > 5998. Por otro lado, el segundo cliente tiene un 79% de probabilidad de no usar la tarjeta y un 21% de probabilidad de que sí lo haga, con un Gini de 0.336. Este cliente sigue la regla: Si TravelSpend es ≥ 1567 e Income es < 59098, FoodSpend > 2711 y TravelSpend ≤ 3008.

Además, se identificaron nodos puros en el modelo, como el nodo (0,47), donde TravelSpend < 1188 y FoodSpend > 4990, y el nodo (1,0), donde TravelSpend < 1188 y FoodSpend > 4972, además del nodo (108,0) mencionado anteriormente. Estos nodos representan decisiones claras y sin ambigüedad sobre el uso o no de la tarjeta de crédito.

